python/cudf/cudf/core/buffer/utils.py

```diff
@@ -134,3 +140,83 @@ def get_spill_lock() -> Union[SpillLock, None]:
     _id = threading.get_ident()
     spill_lock, _ = _thread_spill_locks.get(_id, (None, 0))
     return spill_lock
+
+
+def _clear_property_cache(
+    instance_ref: weakref.ReferenceType[T], nbytes: int, attrname: str
+) -> Optional[int]:
+    """Spill handler that clears the `cached_property` of an instance
+
+    The signature of this function is compatible with SpillManager's
+    register_spill_handler.
+
+    To avoid keeping instance alive, we take a weak reference of the instance.
+
+    Parameters
+    ----------
+    instance_ref
+        Weakref of the instance
+    nbytes : int
+        Size of the cached data
+    attrname : str
+        Name of the cached attribute
+
+    Return
+    ------
+    int
+        Number of bytes cleared
+    """
+
+    instance = instance_ref()
+    if instance is None:
+        return 0
+
+    cached = instance.__dict__.get(attrname, None)
+    if cached is None:
+        return None  # The cached has been cleared
+
+    # If `cached` is known outside of the cache, we cannot free any
+    # memory by clearing the cache. We have three inside references:
+    # `instance.__dict__`, `cached`, and `sys.getrefcount`.
+    if sys.getrefcount(cached) > 3:
+        return None
+
+    instance.__dict__.pop(attrname, None)  # Clear cache atomically
+    return nbytes
+
+
+class cached_property(functools.cached_property):
+    """A version of `cached_property` that delete instead of spill the cache
+
+    When spilling is disabled (the default case), this decorator is identical
+    to `functools.cached_property`.
+
+    When spilling is enabled, this property register a spill handler for
+    the cached data that deletes the data rather than spilling it. For now,
+    only cached Columns are handled this way.
+    See `SpillManager.register_spill_handler`.
+    """
+
+    def __get__(self, instance: T, owner=None):
+        cache_hit = self.attrname in instance.__dict__
+        ret = super().__get__(instance, owner)
+        if cache_hit or not isinstance(ret, cudf.core.column.ColumnBase):
+            return ret
+
+        manager = get_global_manager()
+        if manager is None:
+            return ret
+
+        buf = ret.base_data
```

Comment:
What about the `mask` buffer of the `Column`, or - in the case of nested columns - its children?

Reply:
Good point.  

I have simplified the code a bit. Instead of supporting any type, we now check that the instance is a `RangeIndex`. I haven't found any other use of `cached_property` where the cache can be the sole owner of a buffer, so I think we should limit the scope to `RangeIndex` for now. 

Reply:
Seems related to the question I asked above. I'm OK with not overengineering to support anything else, although it would be good to ensure that if we do need to support other types later it only requires adding new code and not restructuring the existing code (hence my comment above).